go/ethadapter/blockprovider.go

```diff
@@ -0,0 +1,230 @@
+
+const (
+	statusCodeStopped blockProviderStatus = iota
+	statusCodeRunning
+)
+
+const (
+	waitingForBlockTimeout = 30 * time.Second
+)
+
+var one = big.NewInt(1)
+
+func NewEthBlockProvider(ethClient EthClient, logger gethlog.Logger) *EthBlockProvider {
+	return &EthBlockProvider{
+		ethClient:     ethClient,
+		ctx:           context.TODO(),
+		streamCh:      make(chan *types.Block),
+		runningStatus: new(int32),
+		logger:        logger,
+	}
+}
+
+// EthBlockProvider streams blocks from the ethereum L1 client in the order expected by the enclave (consecutive, canonical blocks)
+type EthBlockProvider struct {
+	ethClient EthClient
+	ctx       context.Context
+	logger    gethlog.Logger
+	streamCh  chan *types.Block
+
+	// the state of the block provider
+	currentHead   *types.Header // most recently sent block (reset to nil when a `StartStreaming...` method is called)
+	runningStatus *int32        // 0 = stopped, 1 = running
+}
+
+func (e *EthBlockProvider) start(fromHeight *big.Int) {
+	e.runningStatus = new(int32)
+	go e.streamBlocks(fromHeight)
+}
+
+// StartStreamingFromHash will look up the hash block, find the appropriate height (LCA if there have been forks) and
+// then call StartStreamingFromHeight based on that
+func (e *EthBlockProvider) StartStreamingFromHash(latestHash gethcommon.Hash) (<-chan *types.Block, error) {
+	ancestorBlk, err := e.latestCanonAncestor(latestHash)
+	if err != nil {
+		return nil, err
+	}
+	return e.StartStreamingFromHeight(increment(ancestorBlk.Number()))
+}
+
+// StartStreamingFromHeight will (re)start streaming from the given height, closing out any existing stream channel and
+// returning the fresh channel - the next block will be the requested height
+func (e *EthBlockProvider) StartStreamingFromHeight(height *big.Int) (<-chan *types.Block, error) {
+	// block heights start at 1
+	if height.Cmp(one) < 0 {
+		height = one
+	}
+	if e.streamCh != nil {
+		close(e.streamCh)
+	}
+	e.streamCh = make(chan *types.Block)
+	if e.stopped() {
+		// if the provider is stopped (or not yet started) then we kick off the streaming processes
+		e.start(height)
+	}
+	return e.streamCh, nil
+}
+
+func (e *EthBlockProvider) Stop() {
+	atomic.StoreInt32(e.runningStatus, int32(statusCodeStopped))
+}
+
+func (e *EthBlockProvider) IsLive(h gethcommon.Hash) bool {
+	l1Head, f := e.ethClient.FetchHeadBlock()
+	return f && h == l1Head.Hash()
+}
+
+// streamBlocks is the main loop. It should be run in a separate go routine. It will stream catch-up blocks from requested height until it
+// reaches the latest live block, then it will block until next live block arrives
+// It blocks when:
+// - publishing a block, it blocks on the outbound channel until the block is consumed
+// - awaiting a live block, when consumer is completely up-to-date it waits for a live block to arrive
+func (e *EthBlockProvider) streamBlocks(fromHeight *big.Int) {
+	atomic.StoreInt32(e.runningStatus, int32(statusCodeRunning))
```

Comment:
there is a slight inconsistency here. 
"runningStatus" is being checked in a method but set directly.
I'd say we either wrap all accesses to the accesses in methods or access it directly everywhere.

Reply:
Have switched to using an internal setStatus/getStatus for all of them. And then operating with the enum, hopefully more consistent but open to ideas if not.